src/deluge/gui/views/midi_session_view.cpp

```diff
@@ -0,0 +1,949 @@
+						}
+					}
+					else if (instrument->type == InstrumentType::KIT) {
+						if (!instrumentClipView.getAffectEntire()) {
+							if (patchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+								paramKind = Param::Kind::PATCHED;
+								paramID = patchedParamShortcuts[xDisplay][yDisplay];
+							}
+							else if (unpatchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+								paramKind = Param::Kind::UNPATCHED_SOUND;
+								paramID = unpatchedParamShortcuts[xDisplay][yDisplay];
+							}
+						}
+						else {
+							if (unpatchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+								paramKind = Param::Kind::UNPATCHED_SOUND;
+								paramID = unpatchedParamShortcuts[xDisplay][yDisplay];
+							}
+							else if (globalEffectableParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+								paramKind = Param::Kind::UNPATCHED_GLOBAL;
+								paramID = globalEffectableParamShortcuts[xDisplay][yDisplay];
+							}
+						}
+					}
+					else if (instrument->type == InstrumentType::AUDIO) {
+						if (unpatchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+							paramKind = Param::Kind::UNPATCHED_SOUND;
+							paramID = unpatchedParamShortcuts[xDisplay][yDisplay];
+						}
+						else if (globalEffectableParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+							paramKind = Param::Kind::UNPATCHED_GLOBAL;
+							paramID = globalEffectableParamShortcuts[xDisplay][yDisplay];
+						}
+					}
+					if ((paramKind != Param::Kind::NONE) && (paramID != kNoParamID)) {
+						modelStackWithParam = automationInstrumentClipView.getModelStackWithParam(
+						    modelStackWithTimelineCounter, instrumentClip, paramID, paramKind);
+					}
+				}
+			}
+		}
+
+		if (displayError && (paramToCC[xDisplay][yDisplay] == ccNumber)
+		    && (!modelStackWithParam || !modelStackWithParam->autoParam)) {
+			if (patchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+				paramKind = Param::Kind::PATCHED;
+				paramID = patchedParamShortcuts[xDisplay][yDisplay];
+			}
+			else if (unpatchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+				paramKind = Param::Kind::UNPATCHED_SOUND;
+				paramID = unpatchedParamShortcuts[xDisplay][yDisplay];
+			}
+			else if (globalEffectableParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+				paramKind = Param::Kind::UNPATCHED_GLOBAL;
+				paramID = globalEffectableParamShortcuts[xDisplay][yDisplay];
+			}
+
+			if (display->haveOLED()) {
+				DEF_STACK_STRING_BUF(popupMsg, 40);
+
+				const char* name = getParamDisplayName(paramKind, paramID);
+				if (name != l10n::get(l10n::String::STRING_FOR_NONE)) {
+					popupMsg.append("Can't control: \n");
+					popupMsg.append(name);
+				}
+				display->displayPopup(popupMsg.c_str());
+			}
+			else {
+				display->displayPopup(l10n::get(l10n::String::STRING_FOR_PARAMETER_NOT_APPLICABLE));
+			}
+		}
+	}
+
+	return modelStackWithParam;
+}
+
+void MidiSessionView::updateMappingChangeStatus() {
```

Comment:
add description

it's looking to see if there are any changes to the midi cc mappings since the last time you loaded saved settings